File: src/fsmes/services/reasons.py
"""The downtime vocabulary: what this list is, and what only it knows.

The lifecycle itself - draft, validate, approve, retire, undo, and the rule
that **retiring a code changes what may be chosen next and never what was
chosen before** - lives in `fsmes.services.vocabulary`, where the
non-conformance severities read the same words. It was written here first,
for one table, and moved down when the second vocabulary arrived: two copies
of that rule would have been two copies that drift, and the one that drifts
is the one somebody reads.

What stays here is everything true of the downtime reasons and of no other
list: that what carries a code is a recorded equipment-state interval, that
the machines which have chosen a word are a fact the history already holds,
and that the list's own routes spell `drafts` and `vocabulary`.
"""

from __future__ import annotations

import re

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from fsmes.domain import DowntimeReason, Equipment, EquipmentState
from fsmes.services import vocabulary as lifecycle

#: A code is lowercase, starts with a letter, and holds letters, digits and
#: underscores. It is grouped on, published, and compared across a fleet, so
#: it is spelled the way a topic segment is spelled rather than the way a
#: sentence is.
CODE = re.compile(r"^[a-z][a-z0-9_]{1,39}$")

#: Segments the vocabulary's own routes already spell under
#: `/equipment/downtime-reasons/`. A code that spelled one of these would be
#: unreachable: `/equipment/downtime-reasons/drafts` is the drafts queue, so a
#: reason called `drafts` could never have its history read. Refused at
#: drafting time rather than discovered by a person who cannot open their own
#: reason.
RESERVED = ("drafts", "vocabulary")

#: The statuses a revision can be in when it is the one the plant is currently
#: living with. Exactly one revision of a code is in force at a time.
IN_FORCE = lifecycle.IN_FORCE


def intervals_labelled(session: Session, code: str) -> int:
    """How many equipment-state intervals carry this code, ever."""
    return session.scalar(
        select(func.count()).select_from(EquipmentState)
        .where(EquipmentState.reason_code == code)) or 0


def intervals_by_code(session: Session) -> dict[str, int]:
    """How many recorded intervals carry each code, in one query.

    The screen needs this for every row at once - it is what a person is told
    before they retire a code, rather than after the refusal - and a count per
    code would be one query per reason.
    """
    rows = session.execute(
        select(EquipmentState.reason_code, func.count())
        .where(EquipmentState.reason_code.is_not(None))
        .group_by(EquipmentState.reason_code)).all()
    return {code: count for code, count in rows}


#: What this vocabulary is. Everything generic about it is read from here by
#: `fsmes.services.vocabulary`; what is specific to downtime is the two
#: counters above and the two functions below.
VOCAB = lifecycle.Vocabulary(
    name="downtime_reason",
    noun="downtime reason",
    model=DowntimeReason,
    code=CODE,
    code_rule=(
        "A code is two to forty characters, starts with a lowercase letter, and "
        "holds lowercase letters, digits and underscores - it is grouped on and "
        "published, not a sentence."),
    route="/equipment/downtime-reasons",
    reserved=RESERVED,
    carried_field="labels_intervals",
    counted="recorded interval",
    carried=intervals_labelled,
    carried_by_code=intervals_by_code,
)


# --------------------------------------------------------------- reading it


def revisions(session: Session, code: str) -> list[DowntimeReason]:
    return lifecycle.revisions(session, VOCAB, code)


def in_force(session: Session, code: str) -> DowntimeReason | None:
    """The revision the plant is living with: approved, or retired."""
    return lifecycle.in_force(session, VOCAB, code)


def open_draft(session: Session, code: str) -> DowntimeReason | None:
    """The draft waiting on somebody, if there is one. At most one per code."""
    return lifecycle.open_draft(session, VOCAB, code)


def approved(session: Session) -> list[DowntimeReason]:
    """The vocabulary in force, in code order.

    A retired code is not here - that is what retiring it did - and its
    intervals keep their labels regardless.
    """
    return lifecycle.approved(session, VOCAB)


def catalog(session: Session) -> dict[str, str]:
    """`{code: sentence}` for the approved vocabulary - the shape the screens
    already read `/triggers/catalog` in."""
    return lifecycle.catalog(session, VOCAB)


def names(session: Session) -> dict[str, str]:
    """`{code: name}` for the approved vocabulary - what the button says."""
    return lifecycle.names(session, VOCAB)


def labels(session: Session) -> dict[str, str]:
    """`{code: name}` for every code the plant has ever put in force, retired
    ones included.

    What the analysis reads. A retired code still labels the intervals it
    labelled, so a pareto that only knew the approved list would print the
    bare code for them - the name they were given when somebody chose them is
    the honest label, and it stops changing the moment the code is retired.
    """
    return lifecycle.labels(session, VOCAB)


def drafts(session: Session, limit: int = 50, offset: int = 0
           ) -> tuple[list[DowntimeReason], int]:
    """Every draft waiting on somebody, oldest first, and how many there are."""
    return lifecycle.drafts(session, VOCAB, limit=limit, offset=offset)


def machines_labelling(session: Session, code: str) -> list[tuple[str, int]]:
    """Which machines have recorded a stop under this code, and how many each.

    Most first, then by machine code, so the answer is the same every time it
    is asked. This is what makes a word in the vocabulary point at somewhere
    real: the list itself is plant-wide, but the floor that actually chooses a
    word is a fact the history already holds, and a screen that wants to show
    an approver the effect of a change has to stand somewhere.

    An empty list is an answer, not a gap: a code nobody has ever chosen has
    no machine behind it, and guessing one would be inventing a place.
    """
    rows = session.execute(
        select(Equipment.code, func.count())
        .join(EquipmentState, EquipmentState.equipment_id == Equipment.id)
        .where(EquipmentState.reason_code == code)
        .group_by(Equipment.code)).all()
    return sorted(((machine, count) for machine, count in rows),
                  key=lambda row: (-row[1], row[0]))


def vocabulary(session: Session) -> list[dict]:
    """Every code the plant has ever had, in code order: the revision in force,
    the draft waiting on somebody, and how much history carries the code.

    The catalogue answers *what may an operator choose*, which is the approved
    list and nothing else. This answers *what is the plant's vocabulary*, which
    includes the retired words it still reads in its own history and the drafts
    nobody has signed.
    """
    return lifecycle.vocabulary(session, VOCAB)


# --------------------------------------------------------------- writing it


def define(session: Session, *, code: str, name: str, description: str = "",
           retires: bool = False, labels_intervals: int | None = None,
           actor: str = "system") -> DowntimeReason:
    """Draft a reason: a new code, a change to one, or its retirement.

    One verb, because from the plant's side they are one act - *here is what I
    think the list should say next* - and all three wait for the same person.
    An open draft is edited in place: nobody is choosing from it yet.
    """
    return lifecycle.define(
        session, VOCAB, code=code, name=name, description=description,
        retires=retires, labels_records=labels_intervals, actor=actor)


def approve(session: Session, code: str, revision: int,
            actor: str = "system") -> DowntimeReason:
    """Put one revision in force, superseding whatever the plant had.

    Any revision, not only a draft: approving the previous one is how a
    vocabulary change is undone, and an undo that had to be re-drafted first
    is an undo nobody reaches for in the ten minutes it matters.
    """
    return lifecycle.approve(session, VOCAB, code, revision, actor=actor)


def out(row: DowntimeReason) -> dict:
    return lifecycle.out(VOCAB, row)

File: src/fsmes/web/reasons.js
/* Engineering › Downtime reasons: the plant's own words for why a machine
   stopped.

   PR #85 built this vocabulary end to end for *signing* — the catalogue, the
   station select, the pareto, and the Waiting-for-you panel that puts a draft
   in force. It built nowhere to *draft* one, so the only person who could
   exercise the drafting half was somebody with a terminal and a token. This
   screen is that missing half, and nothing else: it drafts, it edits an open
   draft, and it drafts a retirement. Signing stays where the design put it,
   on the Floor screen's panel, in front of whoever holds `process.approve`.

   Modelled on Engineering › Triggers, which is the same shape of screen — a
   draft-then-approve lifecycle with a form above the list it feeds — so the
   filter bar, the client-side page, the status pills and the count all read
   the same on both. */

const { $, el, api, fmt, toast } = window.FS;

const REFRESH_MS = 15000;
const PAGE = 25;

let rows = [];
let offset = 0;

function live(ok) {
  $("#live-dot").className = "dot" + (ok ? " ok" : " bad");
  $("#live-text").textContent = ok ? "live" : "reconnecting…";
}

function fail(error) {
  live(false);
  const banner = $("#banner");
  banner.textContent = error.message;
  banner.classList.remove("hidden");
}

/* What the plant is living with right now: the approved or retired revision,
   or nothing at all where the word has never been signed. A word can be in
   force AND have a change waiting, and saying only the second of those would
   tell an engineer a word is off the floor while operators are choosing it. */
function inForce(row) {
  return row.in_force ? row.in_force.status : "none";  // approved | retired | none
}

/* The status cell, as one or two pills: where the word stands, and then what
   is waiting on somebody. A retirement waiting is not the same act as a
   rename waiting, and the difference decides whether the word survives. */
function pillsFor(row) {
  if (!row.in_force) {
    return row.draft
      ? [["planned", "drafted, not signed"]]
      : [["down", "never signed"]];
  }
  const where = row.in_force.status === "retired"
    ? ["down", "retired"] : ["running", "on the list"];
  if (!row.draft) return [where];
  return [where, ["planned", row.draft.retires ? "retirement waiting" : "change waiting"]];
}

/* The revision a reader is looking at: what is in force, because that is what
   the operators have; a word with nothing in force shows its draft, which is
   all there is of it. */
function showing(row) {
  return row.in_force || row.draft || null;
}

/* A search reads everything the plant has written about a word, the draft
   included - looking for a reason you half-remember drafting and not finding
   it because it is not signed yet is the search failing. */
function matches(row, q, status) {
  if (status && (status === "draft" ? !row.draft : inForce(row) !== status)) return false;
  if (!q) return true;
  const parts = [row.code];
  for (const rev of [row.in_force, row.draft]) {
    if (rev) parts.push(rev.name || "", rev.description || "");
  }
  return parts.join(" ").toLowerCase().includes(q);
}

async function load() {
  const data = await api("/equipment/downtime-reasons/vocabulary");
  rows = data.reasons;
  draw();
  window.__fsmesPageData = { vocabulary: rows, total: data.total };
}

function draw() {
  const q = $("#v-q").value.trim().toLowerCase();
  const status = $("#v-status").value;
  const matching = rows.filter((row) => matches(row, q, status));
  const page = FS.clientPage(matching, offset, PAGE);
  offset = page.offset;

  const body = $("#reasons-table tbody");
  body.replaceChildren();
  for (const row of page.items) body.append(drawRow(row));

  const empty = $("#v-empty");
  empty.classList.toggle("hidden", page.items.length > 0);
  if (!page.items.length) {
    empty.textContent = rows.length
      ? "No word in this plant's vocabulary matches."
      : "This plant has no downtime vocabulary yet. Until it has one the station "
        + "screen keeps its text box, and the pareto groups on whatever was typed. "
        + "Draft the first word above.";
  }

  // The count says what it covers: the filter's total, and the plant's.
  $("#reason-count").textContent = FS.countText(page, rows.length, "in the vocabulary");
  FS.pager($("#v-pager"), page, (next) => { offset = next; draw(); });

  // The tiles count the plant's vocabulary, never the page's (STYLE.md 5).
  const by = (state) => rows.filter((row) => inForce(row) === state).length;
  $("#kpi-approved").textContent = by("approved").toLocaleString();
  $("#kpi-drafts").textContent = rows.filter((row) => row.draft).length.toLocaleString();
  $("#kpi-retired").textContent = by("retired").toLocaleString();
  $("#kpi-codes").textContent = rows.length.toLocaleString();

  // The form's code box offers what is already there, so drafting a second
  // revision of a word is typing its name rather than remembering its code.
  $("#reason-codes").replaceChildren(...rows.map((row) => {
    const option = document.createElement("option");
    option.value = row.code;
    option.label = (showing(row) || {}).name || "";
    return option;
  }));
}

function drawRow(row) {
  const shown = showing(row) || {};
  const tr = el("tr");

  const what = el("td");
  what.append(el("span", "code", row.code));
  if (shown.name) what.append(el("span", "small", ` ${shown.name}`));
  // What a waiting draft would rename it to, where that is the change: the
  // pill says a change is waiting and this says what the change is.
  if (row.draft && !row.draft.retires && row.in_force
      && row.draft.name !== row.in_force.name) {
    what.append(el("div", "muted small", `waiting: ${row.draft.name}`));
  }
  tr.append(what);

  tr.append(el("td", "small", shown.description || ""));

  const status = el("td");
  for (const [cls, word] of pillsFor(row)) {
    status.append(el("span", `pill ${cls}`, word), " ");
  }
  tr.append(status);

  // The revision in force, and the one waiting behind it where there is one.
  const revision = el("td", "num", shown.revision === undefined ? "—" : String(shown.revision));
  if (row.draft && row.in_force) {
    revision.textContent = `${row.in_force.revision} → ${row.draft.revision}`;
    revision.title = `revision ${row.in_force.revision} is in force; `
      + `revision ${row.draft.revision} is drafted and waiting to be signed`;
  }
  tr.append(revision);

  const drafted = row.draft || shown;
  tr.append(el("td", "muted small", drafted.created_by
    ? `${drafted.created_by}${drafted.on_behalf_of ? ` for ${drafted.on_behalf_of}` : ""} · ${fmt.stamp(drafted.created_at)}`
    : "—"));
  // Unknown renders as a dash, never as a blank that reads like nobody
  // (STYLE.md 8): a draft has not been signed, which is not the same as
  // having been signed by no one.
  tr.append(el("td", "muted small", (row.in_force && row.in_force.approved_by)
    ? `${row.in_force.approved_by} · ${fmt.stamp(row.in_force.approved_at)}`
    : "—"));

  const labels = el("td", "num", row.labels_intervals.toLocaleString());
  labels.title = `${row.labels_intervals.toLocaleString()} recorded intervals carry ${row.code}`;
  tr.append(labels);

  tr.append(actionsFor(row));
  return tr;
}

function actionsFor(row) {
  const cell = el("td");
  if (!FS.can("process.define")) return cell;

  if (row.draft) {
    const edit = el("button", "small", "Edit draft");
    edit.type = "button";
    edit.addEventListener("click", () => editDraft(row));
    cell.append(edit);
  } else {
    const revise = el("button", "small", "New revision");
    revise.type = "button";
    revise.addEventListener("click", () => editDraft(row));
    cell.append(revise);
  }

  // Only a word that is actually on the list can be taken off it; the API
  // says so too, and this is not a second copy of that rule but the reason
  // the button is absent rather than refused.
  if (inForce(row) === "approved" && !(row.draft && row.draft.retires)) {
    const retire = el("button", "ghost small", "Retire");
    retire.type = "button";
    retire.addEventListener("click", () => retireIt(row));
    cell.append(" ", retire);
  }
  return cell;
}

/* ---------- drafting ---------- */

function editDraft(row) {
  // The draft where there is one: that is the text being worked on. A form
  // that loaded the approved wording over the top of an unsigned edit would
  // quietly undo it on save.
  const shown = row.draft || row.in_force || {};
  $("#r-code").value = row.code;
  $("#r-name").value = shown.name || "";
  $("#r-description").value = shown.description || "";
  $("#form-title").textContent = row.draft
    ? `Editing the open draft of ${row.code}`
    : `A new revision of ${row.code}`;
  $("#r-cancel").classList.remove("hidden");
  $("#r-name").focus();
}

function resetForm() {
  $("#reason-form").reset();
  $("#form-title").textContent = "A new word";
  $("#r-cancel").classList.add("hidden");
}

async function submit(event) {
  event.preventDefault();
  try {
    // Nothing here re-states a rule the server owns. The code's spelling, its
    // uniqueness and the words the product has already taken are checked in
    // one place, and what a person reads when one of them refuses is the
    // server's own sentence — a second copy in JavaScript is a second copy
    // that drifts, and the one that drifts is the one people read.
    const out = await api("/equipment/downtime-reasons", { method: "POST", body: {
      code: $("#r-code").value.trim(),
      name: $("#r-name").value.trim(),
      description: $("#r-description").value.trim(),
    } });
    toast(`${out.code} rev ${out.revision} drafted — it reaches the floor when `
      + "somebody signs it on the Floor screen");
    resetForm();
    await load();
  } catch (err) { toast(err.message, "bad"); }
}

async function retireIt(row) {
  // The number comes before the decision, not after a refusal. The API asks
  // the drafter to say how much history carries the code; this screen is
  // where they are told what that number is.
  const carried = row.labels_intervals;
  const name = (row.in_force || {}).name || row.code;
  const ask = carried
    ? `Retire ${row.code}? ${carried.toLocaleString()} recorded `
      + `${carried === 1 ? "interval carries" : "intervals carry"} it. They keep `
      + "the label — retiring changes what may be chosen next, never what was "
      + "chosen before. The retirement is a draft until somebody signs it."
    : `Retire ${row.code}? No recorded interval carries it. The retirement is a `
      + "draft until somebody signs it.";
  if (!window.confirm(ask)) return;

  try {
    await api("/equipment/downtime-reasons", { method: "POST", body: {
      code: row.code, name, description: (row.in_force || {}).description || "",
      retires: true, labels_intervals: carried,
    } });
    toast(`${row.code}: retirement drafted — it leaves the list when somebody signs it`);
    await load();
  } catch (err) { toast(err.message, "bad"); }
}

/* ---------- boot ---------- */

(async function boot() {
  await FS.whoami().catch(() => {});
  FS.applyCapGates();
  $("#reason-form").addEventListener("submit", submit);
  $("#r-cancel").addEventListener("click", resetForm);
  for (const id of ["v-q", "v-status"]) {
    $(`#${id}`).addEventListener("input", () => { offset = 0; draw(); });
  }
  const run = () => load().then(() => live(true)).catch(fail);
  await run();
  setInterval(run, REFRESH_MS);
})().catch(fail);
